rpmbuild.c

```diff
@@ -337,13 +337,17 @@ static int isSpecFile(const char * specfile)
 static char * getTarSpec(const char *arg)
 {
     char *specFile = NULL;
+    char *tmpSpecFile = NULL;
+    char *specDir;
+    char *specBase;
     const char **spec;
     char tarbuf[BUFSIZ];
-    int gotspec = 0;
+    int gotspec = 0, res;
     FD_t fd = NULL;
     static const char *tryspec[] = { "Specfile", "\\*.spec", NULL };
 
-    if (!(fd = rpmMkTempFile(NULL, &specFile)))
+    specDir = rpmGetPath("%{_tmppath}", NULL);
+    if (!(fd = rpmMkTempFile(NULL, &tmpSpecFile)))
```

Comment:
I'm thinking if it makes sense to create this as a temp file, only to be renamed to the real thing in the end? :)

Originally, I suppose, it was done because we didn't want to write the final name into `_specdir` before being done with it (as to not risk leaving an incomplete file behind in case of a failure), however now that we're keeping the file in `_tmpdir` even after the rename, we could perhaps just use the final name?

Reply:
My main issue with this patch is that it tries to place a "predictable" filename into a world-writable directory. This is always gets people going about security, but it also can cause unpredictable failures for no good reason (two users trying to do the same thing at the same time, stranger things have happened). That is a no-go.

So we kinda do need a temporary file for this, but we also need a place to put the final spec. The specdir as used by the original code wasn't optimal either as you could end up overwriting your own data with zero warning (has actually happened to me). One "easy" solution would be just using current directory, but with similar risks of overwriting data. So to handle this in a way that cannot randomly fail, or overwrite your data, if you happen to have a file by the same name around, is to create a temporary *directory* and place the spec (and anything else we might need) there.

Reply:
...

The other, simpler option would be going back to creating %_specdir if it doesn't exist. That was not the main target in the 4.16 change anyway, %_sourcedir was.

Reply:
Hmm, using %_tmpdir brings on yet another problem: rename() will then fail "randomly" depending on your partitioning scheme and commonly fail because of it, which is also not okay. For such a silly little thing this is annoyingly complicated and subtle...